pydantic/main.py

```diff
@@ -640,15 +649,18 @@ def _iter(
 
         for k, v in self.__dict__.items():
             if allowed_keys is None or k in allowed_keys:
-                yield k, self._get_value(
+                value = self._get_value(
                     v,
                     to_dict=to_dict,
                     by_alias=by_alias,
                     include=value_include and value_include.for_element(k),
                     exclude=value_exclude and value_exclude.for_element(k),
                     exclude_unset=exclude_unset,
                     exclude_defaults=exclude_defaults,
+                    exclude_none=exclude_none,
                 )
+                if (exclude_none and value is not None) or not exclude_none:
```

Comment:
```suggestion
                if not (exclude_none and value is None):
```
I think slightly clearer and more efficient.